### PR TITLE
CommandBox 3.9.0

### PR DESCRIPTION
**Pull request by bdw429s**
CommandBox 3.9.0

### CODE REVIEW

**Comment by ilovezfs** [comment]
it's passing CI with Java 9. What's the problem?

File: Formula/commandbox.rb [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
   homepage "https://www.ortussolutions.com/products/commandbox"
-  url "https://downloads.ortussolutions.com/ortussolutions/commandbox/3.8.0/commandbox-bin-3.8.0.zip"
-  sha256 "e184e380713b6e4e94c1266bf4abc352da2e5c7b19bdb9d5ce5f36665d1c58ec"
+  url "https://downloads.ortussolutions.com/ortussolutions/commandbox/3.9.0/commandbox-bin-3.9.0.zip"
+  sha256 "9497a7db8879150548b05d28594cc7d316556fc3a2e4602d3795b589ba85105e"
 
   bottle :unneeded
 
-  depends_on :java => "1.7+"
+  depends_on :java => "1.7+ < 9"
```
Comment (ilovezfs) [inline Formula/commandbox.rb]: you'd need to require Java 8 specifically and use a wrapper script

**Comment by bdw429s** [comment]
> it's passing CI with Java 9. 
@ilovezfs Not from what I saw.  Check the builds in the first commit (which failed)  My program is not compatible with Java 9.  I couldn't find the valid syntax to only allow java 7 and 8 in the docs.  I need CommandBox to not run on Java 9 or it's going to error out right now.

File: Formula/commandbox.rb (hunk 1 above)
Comment (bdw429s) [reply Formula/commandbox.rb]: @ilovezfs Can you explain what a "wrapper script" means?  I just need to allow either Java 7 or 8, but not 9.  I looked and looked on Google but couldn't find the correct syntax for this.  I couldn't even locate the Homebrew docs that covered the valid expressions above.  I just guessed with what I typed and it seemed to make the builds pass.

**Comment by ilovezfs** [comment]
Ah, it passed because an unsatisfied requirement is an automatic pass.
```
==> SKIPPING: commandbox
Java 1.7+ < 9 is required to install this formula.
JavaRequirement unsatisfied!
You can install with Homebrew-Cask:
 brew cask install caskroom/versions/java8
You can download from:
 https://www.oracle.com/technetwork/java/javase/downloads/index.html
```

**Comment by ilovezfs** [comment]
```
 I couldn't find the valid syntax to only allow java 7 and 8 in the docs.
```
depends_on :java => "x"
```
passes `x` to `/usr/libexec/java_home -v`

```
iMac-TMP:~ joe$ ls -1 /Library/Java/JavaVirtualMachines/
jdk-9.0.1.jdk
jdk1.8.0_151.jdk
zulu-7.jdk
zulu1.7.0_154.jdk
iMac-TMP:~ joe$ /usr/libexec/java_home -v 1.7+
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 1.8+
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 9.0+
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 9.0
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 8.0
Unable to find any JVMs matching version "8.0".
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 1.8
/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 1.7
/Library/Java/JavaVirtualMachines/zulu-7.jdk/Contents/Home
iMac-TMP:~ joe$
```
So you need
```
depends_on :java => "1.8"
```

**Comment by bdw429s** [comment]
> depends_on :java => "1.8"

@ilovezfs But CommandBox runs on Java 7 as well as I stated above.  Wouldn't the syntax you showed above only work on Java 8 or am I missing something?

**Comment by ilovezfs** [comment]
You're not missing anything. To exclude Java 9 we have to pin to Java 8 specifically.

**Comment by bdw429s** [comment]
> depends_on :java => "x"

:point_up:  That doesn't really tell me anything.  Is that a semantic versioning syntax?  Can you use greater than, less than?  Can you have a list of versions?  A range of versions?  All these questions and more were unanswered when I Googled and read through your docs.

**Comment by bdw429s** [comment]
> To exclude Java 9 we have to pin to Java 8 specifically.

Hmm, that certainly feels wrong to say that my application "depends on Java 8" when it doesn't (Since java 7 is also valid) but I'll make that change.

**Comment by ilovezfs** [comment]
`man java_home` the docs are terrible :)

**Comment by ilovezfs** [comment]
In terms of the wrapper script.

```
    libexec.install "box"
    (bin/"box").write_env_script libexec/"box", Language::Java.java_home_env("1.8")
```

**Comment by ilovezfs** [comment]
which results in
```
iMac-TMP:~ joe$ cat /usr/local/Cellar/commandbox/3.9.0/bin/box 
#!/bin/bash
JAVA_HOME="$(/usr/libexec/java_home --version 1.8)" exec "/usr/local/Cellar/commandbox/3.9.0/libexec/box" "$@"
iMac-TMP:~ joe$ 
```

**Comment by ilovezfs** [comment]
@bdw429s as you can see it fails without the wrapper script.

**Comment by jclausen** [comment]
@ilovezfs : Can you clarify on the way to implement the wrapper within the formula? I'm unable to find any documentation in the Formula cookbook.  For clarification on who's who here, @bdw429s is the developer of the binary and  I'm the original submitter of the formula.

**Comment by bdw429s** [comment]
Thanks for the additional help @ilovezfs .  Like Jon, I'm still a little unsure of exactly what to do here.  
Am I just replacing these lines in my rb file:

```
  def install
    bin.install "box"
    doc.install resource("apidocs")
  end
```

with this?

```
  def install
   libexec.install "box"
    (bin/"box").write_env_script libexec/"box", Language::Java.java_home_env("1.8")
    doc.install resource("apidocs")
  end
```

**Comment by ilovezfs** [comment]
>Am I just replacing these lines in my rb file:

yup.

**Comment by bdw429s** [comment]
Thanks again for the help @ilovezfs Now that I added in the wrapper script (and fixed some indentation) it seems we've appeased the build gods.  Let me know if anything else needs done.

**Comment by ilovezfs** [comment]
You're welcome :)

Shipped! Thanks for the upgrade @bdw429s! :rocket: 

And thanks for pitching in @jclausen!

**Pull request merged**